extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -335,170 +563,210 @@ bool TailFile::storeState(const std::shared_ptr<core::ProcessContext>& context)
+      if (mtime >= std::chrono::time_point_cast<std::chrono::seconds>(state.last_read_time_)) {
+        logger_->log_debug("File %s has mtime >= last read time, so we are going to read it", file_name);
+        matched_files_with_mtime.emplace_back(TailState{path, file_name}, mtime);
+      }
+    }
+    return true;
+  };
+
+  utils::file::FileUtils::list_dir(state.path_, collect_matching_files, logger_, false);
+
+  std::sort(matched_files_with_mtime.begin(), matched_files_with_mtime.end(), [](const TailStateWithMtime &left, const TailStateWithMtime &right) {
+    return std::tie(left.mtime_, left.tail_state_.file_name_) <
+           std::tie(right.mtime_, right.tail_state_.file_name_);
+  });
+
+  if (!matched_files_with_mtime.empty() && state.position_ > 0) {
+    TailState &first_rotated_file = matched_files_with_mtime[0].tail_state_;
+    std::string full_file_name = first_rotated_file.fileNameWithPath();
+    if (utils::file::FileUtils::file_size(full_file_name) >= state.position_) {
+      uint64_t checksum = utils::file::FileUtils::computeChecksum(full_file_name, state.position_);
+      if (checksum == state.checksum_) {
+        first_rotated_file.position_ = state.position_;
+        first_rotated_file.checksum_ = state.checksum_;
       }
-      TailMatchedFileItem item;
-      item.fileName = filename;
-      item.modifiedTime = ((uint64_t) (sb.st_mtime) * 1000);
-      matchedFiles.push_back(item);
     }
   }
-  return true;};
-
-    utils::file::FileUtils::list_dir(file.path_, lambda, logger_, false);
 
-    if (matchedFiles.size() < 1) {
-      logger_->log_debug("No newer files found in directory!");
-      return;
-    }
+  std::vector<TailState> matched_files;
+  std::transform(matched_files_with_mtime.begin(), matched_files_with_mtime.end(), std::back_inserter(matched_files),
+                 [](TailStateWithMtime &tail_state_with_mtime) { return std::move(tail_state_with_mtime.tail_state_); });
+  return matched_files;
+}
 
-    // Sort the list based on modified time
-    std::sort(matchedFiles.begin(), matchedFiles.end(), sortTailMatchedFileItem);
-    TailMatchedFileItem item = matchedFiles[0];
-    logger_->log_info("TailFile File Roll Over from %s to %s", file.current_file_name_, item.fileName);
+void TailFile::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  std::lock_guard<std::mutex> tail_lock(tail_file_mutex_);
 
-    // Going ahead in the file rolled over
-    if (file.current_file_name_ != base_file_name) {
-      logger_->log_debug("Resetting posotion since %s != %s", base_file_name, file.current_file_name_);
-      file.currentTailFilePosition_ = 0;
+  if (tail_mode_ == Mode::MULTIPLE) {
+    if (last_multifile_lookup_ + lookup_frequency_ < std::chrono::steady_clock::now()) {
+      logger_->log_debug("Lookup frequency %d ms have elapsed, doing new multifile lookup", lookup_frequency_.count());
```

Comment:
Printing chrono underlying types for the correct format specifier is tricky. Please see my comment here, a previous similar case: https://github.com/apache/nifi-minifi-cpp/pull/741#discussion_r385725765

Suggestion: wrap the count call in `uint64_t{}` and use `PRIu64` as the format specifier.

Reply:
The underlying types of `seconds` and `milliseconds` are "signed integer type of at least 35 bits" and "signed integer type of at least 45 bits", respectively, so I changed this (and similar logs) to `int64_t{some_duration.count()}` and `PRId64`.

Reply:
You're right, thanks.